### PR TITLE
IOperation API extensions to expose low-level operations

### PR DESCRIPTION
**Pull request by edgardozoppi**
Since there are several requested changes I have decided to create a new PR #21810 with those changes. 
Please take a look to the new PR #21810 instead.

These changes add a new overload to the `SemanticModel.GetOperation` method to generate an IOperation tree from the compiler's internal BoundTree after performing the local rewriting transformation phase. The new low-level IOperation tree reuses most of the original IOperation interfaces and extend them by adding a few newly created ones that can only appear after the local rewriting phase. This has been done to support both C# and VB.

**Remarks**
These changes are experimental so we still need to keep them behind the IOperation feature flag even after releasing the API in Dev 15.5.

More detailed information of the changes can be found in this issue #21507.

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
> and extend them by adding a few newly created ones that can only appear after the local rewriting phase. 

Can you give a rundown on what these are?  Thanks!

**Comment by CyrusNajmabadi** [comment]
> after performing the local rewriting transformation phase.

Can you give a brief explanation of what phases of lowering this does?  i.e. what happens to lambdas, iterators, async, other-code after the "local rewriting" phase.

Thanks!

**Comment by edgardozoppi** [comment]
> Can you give a rundown on what these are? Thanks!

Please take a look to this issue #21507.

> Can you give a brief explanation of what phases of lowering this does? i.e. what happens to lambdas, iterators, async, other-code after the "local rewriting" phase.

This new GetOperation method overload only performs the local rewriting lowering phase by using the `LocalRewriter` classes (C# and VB versions respectively). Not sure exactly what are all the languages constructs that this phase transforms. I guess that is a question that someone from the compiler team can answer more precisely. You can take a look to the `LocalRewriter` implementation for C# [here](http://source.roslyn.io/#Microsoft.CodeAnalysis.CSharp/Lowering/LocalRewriter/LocalRewriter.cs) and for VB [here](http://source.roslyn.io/#Microsoft.CodeAnalysis.VisualBasic/Lowering/LocalRewriter/LocalRewriter.vb).

There are other lowering phases to transform local functions, lambdas, iterators, async/await, etc. All those phases happen after the local rewriting when the compiler has to emit code. But for generating the low-level IOperation tree we are only using the first phase.

File: src/Compilers/CSharp/Portable/Compilation/SyntaxTreeSemanticModel.cs
Comment (agocke) [inline src/Compilers/CSharp/Portable/Compilation/SyntaxTreeSemanticModel.cs]: `return model?.GetOperationWorker(method, node, cancellationToken)`

File: src/Compilers/Core/Portable/Operations/IConditionalGotoStatement.cs
Comment (agocke) [inline src/Compilers/Core/Portable/Operations/IConditionalGotoStatement.cs]: This is unclear. What does it do if this is `false` and the condition is true?

File: src/Compilers/Core/Portable/Generated/Operations.xml.Generated.cs
Comment (agocke) [inline src/Compilers/Core/Portable/Generated/Operations.xml.Generated.cs]: I see generated code here, but I don't see any changes to a generate base.

**Review by agocke — CHANGES_REQUESTED** [review]
I would expect some unit test coverage for these changes before merging.

File: src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs
Comment (mavasani) [inline src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs]: Can we consider changing the API to take only one of the symbol OR body parameters? Alternatively, please file a separate issue to track removal of one of these parameters.

File: src/Compilers/Core/Portable/Generated/Operations.xml.Generated.cs
Comment (mavasani) [reply src/Compilers/Core/Portable/Generated/Operations.xml.Generated.cs]: @agocke even though file is .generated.cs, it is a hand crafted file. https://github.com/dotnet/roslyn/issues/20204 tracks writing an operation generator similar to the syntax and bound node generators.

**Review by mavasani — APPROVED** [review]
LGTM, modulo addition of unit tests. @edgardozoppi Lets work today on adding unit tests for the new APIs.

File: src/Compilers/Core/Portable/Generated/Operations.xml.Generated.cs
Comment (agocke) [reply src/Compilers/Core/Portable/Generated/Operations.xml.Generated.cs]: @mavasani This is very confusing. Can we please rename it to not included `generated` in the file name?

File: src/Compilers/Core/Portable/Operations/IConditionalGotoStatement.cs
Comment (edgardozoppi) [reply src/Compilers/Core/Portable/Operations/IConditionalGotoStatement.cs]: I guess it won't jump in that case. I just copied that property from BoundConditionalGoto.JumpIfTrue property. I think it means if the conditional goto branch is behaving like JE (jumpe on equals) or JNE (jump on not equals). So, if the property JumpIfTrue is false, it will only jump when the condition is false.

File: src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs
Comment (edgardozoppi) [reply src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs]: Done.

File: src/Compilers/Core/Portable/Compilation/SemanticModel.cs
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Compilation/SemanticModel.cs]: >public IOperation GetOperation(IMethodSymbol method, CancellationToken cancellationToken = default) [](start = 8, length = 99)

It is not clear why this API is on SemanticModel. SemanticModel is created for a particular syntax tree, this API doesn't depend on syntax at all. Consider moving it to Compilation.

File: src/Compilers/Core/Portable/Operations/ISequenceExpression.cs
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Operations/ISequenceExpression.cs]: >Expressions [](start = 35, length = 11)

SideEffects?

File: src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs]: >(MethodSymbol)method [](start = 27, length = 20)

Please use EnsureCSharpSymbolOrNull helper.

File: src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs]: >CheckSymbol(csmethod); [](start = 12, length = 22)

It looks like this helper is using "symbol" as parameter name, I would expect "method" to be used for this call site instead.

File: src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs]: >CheckSyntaxNode(csbody); [](start = 12, length = 24)

I think consumer might be confused by an exception thrown by this helper because syntax is not an input for the GetOperation API.

File: src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs]: >DiagnosticBag.GetInstance() [](start = 30, length = 27)

Are we leaking diagnostics bag from the pool?

File: src/Compilers/CSharp/Portable/Symbols/MethodSymbolExtensions.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/MethodSymbolExtensions.cs]: >var [](start = 12, length = 3)

Please spell out the type.

File: src/Compilers/CSharp/Portable/Symbols/MethodSymbolExtensions.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/MethodSymbolExtensions.cs]: >syntaxReference [](start = 23, length = 15)

Could ```syntaxReference``` be null here? Would it be safe to call GetCSharpSyntax in this case?

File: src/Compilers/CSharp/Portable/Symbols/MethodSymbolExtensions.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/MethodSymbolExtensions.cs]: >FirstOrDefault [](start = 67, length = 14)

It is not clear why FirstOrDefault is the right thing to do. It is not clear why are we doing this through DeclaringSyntaxReferences at all. I think that we should only handle top-level methods explicitly declared in source (not lambdas, not local functions, etc.) It'll probably make sense to to do what MethodCompiler.BindMethodBody does rather than inventing some new way of getting the bound body, perhaps share some code with that method.

File: src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs]: >result = _operationFactory.Create(loweredBody); [](start = 16, length = 47)

It feels wrong to tie lowered tree to any SemanticModel. SemanticModel deals with un-lowered bound trees, who knows what negative effects will this operation have, how will it affect the caches, etc. I think there should be no interaction between SemanticModel and the tree built of off lowered bound nodes.

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: >s.Kind != OperationKind.None [](start = 136, length = 28)

Why is this the right condition? Isn't this potentially going to remove significant portion of the tree?

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: >s.Kind != OperationKind.None [](start = 132, length = 28)

Same question.

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IConditionalGotoStatement.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IConditionalGotoStatement.cs]: >IConditionalGotoStatement (JumpIfTrue: False) (OperationKind.ConditionalGotoStatement) (Syntax: 'p < 0') [](start = 4, length = 104)

From this output, how do I make sense where the jump is going to?

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IConditionalGotoStatement.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IConditionalGotoStatement.cs]: >IBranchStatement (BranchKind.GoTo) (OperationKind.BranchStatement) (Syntax: 'if (p < 0)  ... else p = 1;') [](start = 4, length = 106)

How can I determine where this branch jumps to?

**Comment by AlekseyTs** [comment]
It looks like added tests do not cover all added/modified code paths.

**Comment by AlekseyTs** [comment]
Done with review pass for C# side of changes (iteration 12).

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IConditionalGotoStatement.cs
Comment (edgardozoppi) [reply src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IConditionalGotoStatement.cs]: The problem is that the labels of IConditionalGotoStatement are automatically generated by the compiler when lowering the method body, so they have different names depending on the order in which the unit tests run.

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IConditionalGotoStatement.cs
Comment (edgardozoppi) [reply src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IConditionalGotoStatement.cs]: Same as before.

**Comment by edgardozoppi** [comment]
Since there are several requested changes I have decided to create a new PR #21810 with those changes. Also, because I did the changes in a new branch in my fork that has all the latest updates of the `features/ioperation` branch.

Please continue the review in the new PR #21810.

**Pull request closed without merging**
